toolchain/check/context.cpp

```diff
@@ -362,200 +371,552 @@ class Context::PendingBlock {
+// to be materialized.
+static bool IsValidExpressionCategoryForConversionTarget(
+    SemIR::ExpressionCategory category,
+    Context::ConversionTarget::Kind target_kind) {
+  switch (target_kind) {
+    case Context::ConversionTarget::Value:
+      return category == SemIR::ExpressionCategory::Value;
+    case Context::ConversionTarget::ValueOrReference:
+    case Context::ConversionTarget::Discarded:
+      return category == SemIR::ExpressionCategory::Value ||
+             category == SemIR::ExpressionCategory::DurableReference ||
+             category == SemIR::ExpressionCategory::EphemeralReference ||
+             category == SemIR::ExpressionCategory::Initializing;
+    case Context::ConversionTarget::Initializer:
+    case Context::ConversionTarget::FullInitializer:
+      return category == SemIR::ExpressionCategory::Initializing;
   }
 }
 
-auto Context::InitializeAndFinalize(Parse::Node parse_node,
-                                    SemIR::NodeId target_id,
-                                    PendingBlock& target_block,
-                                    SemIR::NodeId value_id) -> SemIR::NodeId {
-  auto init_id = InitializeImpl(parse_node, target_id, target_block, value_id);
-  if (init_id == SemIR::NodeId::BuiltinError) {
-    return init_id;
+auto Context::PerformBuiltinConversion(Parse::Node parse_node,
+                                       SemIR::NodeId value_id,
+                                       ConversionTarget target)
+    -> SemIR::NodeId {
+  auto value = semantics_ir().GetNode(value_id);
+  auto value_type_id = value.type_id();
+  auto target_type_node = semantics_ir_->GetNode(
+      semantics_ir().GetTypeAllowBuiltinTypes(target.type_id));
+
+  // Various forms of implicit conversion are supported as builtin conversions
+  // rather than being implemented purely in the Carbon prelude as `impl`s of
+  // `ImplicitAs`. There are a few reasons we need to perform some of these
+  // conversions as builtins:
+  //
+  // 1) Conversions from struct and tuple *literals* have special rules, and
+  //    cannot be performed in the library. Specifically, we must recurse into
+  //    the elements of the literal before performing initialization in order
+  //    to avoid unnecessary conversions between expression categories.
+  // 2) (Not implemented yet) Conversion of a facet to a facet type depends on
+  //    the value of the facet, not only its type.
+  // 3) Some of these conversions are used while checking the library
+  //    definition of `ImplicitAs` itself or implementations of it.
+  //
+  // We also expect to see better performance by avoiding an `impl` lookup for
+  // common conversions.
+  //
+  // TODO: We should provide a debugging flag to turn off as many of these
+  // builtin conversions as we can so that we can test that they do the same
+  // thing as the library implementations.
+  //
+  // These builtin conversions all correspond to `final impl`s in the library,
```

Comment:
This sentence seems to contradict some of the above comments about not being able to implement some of these in the library?

Reply:
Hm, I see your point. It's not that these conversions won't exist in the library -- they (eventually) will do -- it's that calling `ImplicitAs` won't do the right thing because it doesn't support taking (eg) a tuple literal with mixed expression categories for its elements.

I've updated the text above to try to not suggest these conversions won't exist in the library.